activesupport/lib/active_support/hash_with_indifferent_access.rb

```diff
@@ -335,16 +335,29 @@ def transform_values(&block)
       dup.tap { |hash| hash.transform_values!(&block) }
     end
 
-    def transform_keys(*args, &block)
-      return to_enum(:transform_keys) unless block_given?
-      dup.tap { |hash| hash.transform_keys!(*args, &block) }
+    NOT_GIVEN = Object.new # :nodoc:
+
+    def transform_keys(hash = NOT_GIVEN, &block)
+      return to_enum(:transform_keys) if (hash == NOT_GIVEN) && !block_given?
+      dup.tap { |h| h.transform_keys!(hash, &block) }
     end
 
-    def transform_keys!
-      return enum_for(:transform_keys!) { size } unless block_given?
-      keys.each do |key|
-        self[yield(key)] = delete(key)
+    def transform_keys!(hash = NOT_GIVEN, &block)
+      if hash.nil?
+        super
```

Comment:
I think the right thing to do would be something like:

```ruby
if NOT_GIVEN.equal?(hash)
  # do the purely block based thing
else
  hash = Hash.try_convert(hash) or super # super will raise the proper error
end
```

I also just looked at the upstream implementation and just noticed that we have another incompatibility with how Hash behaves. It accepts both an argument and a block:

```ruby
>> {foo: 1, bar: 2}.transform_keys(foo: "oof") { |k| k.upcase }
=> {"oof"=>1, :BAR=>2}
```

Reply:
Ooh, I thought I did check that, but you're right, that seems to be how the Ruby version works. However, passing in both Hash and block is indeed not tested nor documented, so it might be just an undefined behavior. Also, that result looks a bit tricky to me. If I'm asked to use both Hash argument and block, I would make it return `{"OOF"=>1, :BAR=>2}` in that case...
Anyway, I'll take a look at this again later.

Reply:
> not tested nor documented

I think it is:

> Returns a new Hash object; each entry has:

>    A key provided by the block.

>    The value from self.

> An optional hash argument can be provided to map keys to new keys. **Any key not given will be mapped using the provided block**, or remain the same if no block is given.

Implementation wise it might be done as:

```ruby
self[hash.fetch(key, &block)] = delete(key)
```

For the concise way.

Reply:
Added the Hash argument + block pattern. The code became a little bit more complex than I imagined, but it covers all patterns now.

Reply:
> [T]hat result looks a bit tricky to me. If I'm asked to use both Hash argument and block, I would make it return `{"OOF"=>1, :BAR=>2}`

For `Hash#transform_keys` in Ruby, the whole point of allowing a hash argument was to allow irregularities. Suppose you wanted to attach `"oo"` to all keys except for key `"a"`, i.e., `{"a" => 1, "b" => 2, "c" => 3, ...}` => `{"a" => 1, "boo" => 2, "coo" => 3, ...}`. If you write the block so as to attach `"foo"` to the keys, and that applied to all keys including `"a"`, it would be difficult to handle the irregularity. Cf. https://bugs.ruby-lang.org/issues/16274#note-13

Thanks for implementing it in this way in the end.